### ISSUE TITLE
Upstream tasks issue with Task.map

### ISSUE THREAD

**Issue by jacques-** [PrefectHQ/prefect#2752]
## Description

I'm encountering a bug that feels related to #2430 that was fixed in master. When an  `upstream_tasks` parameter is provided to a Task.map function and the upstream task errors I get the following:

`TypeError: Cannot map over unsubscriptable object of type <class 'prefect.engine.signals.TRIGGERFAIL'>: TRIGGERFAI...`

There is no issue as long as the upstream task does not fail.

## Expected Behavior

Expected the map function to behave similar to an unmapped function in this case.

## Reproduction

```
from typing import List, Dict
from prefect import Flow
from prefect import Task
from prefect.triggers import any_failed


class TaskA(Task):
    def run(self) -> List[Dict]:
        records = [
            {"data": "dummy"},
            {"data": "dummier"},
            {"data": "dummiest"},
        ]
        return records


class TaskB(Task):
    def run(self) -> None:
        print("doing something unrelated")


class TaskC(Task):
    def run(self, data: List) -> None:
        raise RuntimeError("Uncomment this to fix")
        print(f"Data: {data}")


class TaskD(Task):
    def run(self, data: List) -> None:
        print("Cleaning up failure")


task_a = TaskA()
task_b = TaskB()
task_c = TaskC()
task_d = TaskD(trigger=any_failed)

# Create dummy flow
with Flow("dummy_flow") as flow:
    data = task_a()
    task_b_slug = task_b().slug
    task_c_slug = task_c.map(
        data=data, upstream_tasks=[flow.get_tasks(slug=task_b_slug)]
    ).slug
    task_d.map(data=data, upstream_tasks=[flow.get_tasks(slug=task_c_slug)])

flow.run()
```

## Environment

I tested this in 0.11.5 and master and seems to be the same bahaviour.

**Comment by cicdw** [comment]
Hi @jacques- - what do you expect the behavior to be here?  If an upstream task fails, what should be mapped over?

**Comment by cicdw** [comment]
Ah @jacques- I think the issue here is the way you're referencing tasks.  Note that `flow.get_tasks(slug=task_b_slug)` returns a _list_, so that's where the `List` task is coming from in your logs.  

I recommend referencing tasks more explicitly by changing your code to the following:

```python
with Flow("dummy_flow") as flow:
    data = task_a()
    task_b_copy = task_b()
    task_c_copy = task_c.map(
        data=data, upstream_tasks=[task_b_copy]
    )
    task_d.map(data=data, upstream_tasks=[task_c_copy])
```

You can simplify this further by avoiding the unnecessary copies of task A and task B:

```python
with Flow("dummy_flow") as flow:
    task_c_ref = task_c.map(
        data=task_a, upstream_tasks=[task_b]
    )
    task_d.map(data=task_a, upstream_tasks=[task_c_ref])
```

Both of those code blocks resulted in successful flow runs for me.

**Comment by jacques-** [comment]
Cheers, didn't realise that `get_tasks` was returning a `List`, thanks!

However, I'm still getting failures with your changes in 0.11.5 and master.

Just so we are 100% on the same page, here is the code I'm running:

```
from typing import List, Dict
import logging
import prefect
from prefect import Flow
from prefect import Task
from prefect.triggers import any_failed

prefect_logger = prefect.utilities.logging.get_logger()
prefect_logger.setLevel(logging.INFO)


class TaskA(Task):
    def run(self) -> List[Dict]:
        print("Starting A")
        records = [
            {"data": "dummy"},
            {"data": "dummier"},
            {"data": "dummiest"},
        ]
        return records


class TaskB(Task):
    def run(self) -> None:
        print("Starting B")
        print("doing something unrelated")


class TaskC(Task):
    def run(self, data: List) -> None:
        print("Starting C")
        # raise RuntimeError("Uncomment this to fix")
        print(f"Data: {data}")


class TaskD(Task):
    def run(self, data: List) -> None:
        print("Starting D")
        print("Cleaning up failure")


task_a = TaskA()
task_b = TaskB()
task_c = TaskC()
task_d = TaskD(trigger=any_failed)


with Flow("dummy_flow") as flow:
    task_c_ref = task_c.map(data=task_a, upstream_tasks=[task_b])
    task_d.map(data=task_a, upstream_tasks=[task_c_ref])

flow.run()
```
And the logs I get for 0.11.5:

```
[2020-06-12 06:23:05] INFO - prefect.FlowRunner | Beginning Flow run for 'dummy_flow'
[2020-06-12 06:23:05] INFO - prefect.FlowRunner | Starting flow run.
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskB': Starting task run...
Starting B
doing something unrelated
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskB': finished task run for task with final state: 'Success'
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskA': Starting task run...
Starting A
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskA': finished task run for task with final state: 'Success'
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskC': Starting task run...
[2020-06-12 06:23:05] ERROR - prefect.TaskRunner | Task 'TaskC': unexpected error while running task: TypeError("Cannot map over unsubscriptable object of type <class 'NoneType'>: None...")
Traceback (most recent call last):
  File "/XXX/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/prefect/engine/task_runner.py", line 279, in run
    executor=executor,
  File "/XXX/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/prefect/engine/task_runner.py", line 789, in run_mapped_task
    preview=repr(upstream_state.result)[:10],
TypeError: Cannot map over unsubscriptable object of type <class 'NoneType'>: None...
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskC': finished task run for task with final state: 'Failed'
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskD': Starting task run...
[2020-06-12 06:23:05] ERROR - prefect.TaskRunner | Task 'TaskD': unexpected error while running task: TypeError("Cannot map over unsubscriptable object of type <class 'TypeError'>: TypeError(...")
Traceback (most recent call last):
  File "/XXX/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/prefect/engine/task_runner.py", line 279, in run
    executor=executor,
  File "/XXX/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/prefect/engine/task_runner.py", line 789, in run_mapped_task
    preview=repr(upstream_state.result)[:10],
TypeError: Cannot map over unsubscriptable object of type <class 'TypeError'>: TypeError(...
[2020-06-12 06:23:05] INFO - prefect.TaskRunner | Task 'TaskD': finished task run for task with final state: 'Failed'
[2020-06-12 06:23:05] INFO - prefect.FlowRunner | Flow run FAILED: some reference tasks failed.
```

And for master:

```
[2020-06-12 06:19:52] INFO - prefect.FlowRunner | Beginning Flow run for 'dummy_flow'
[2020-06-12 06:19:52] INFO - prefect.FlowRunner | Starting flow run.
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskA': Starting task run...
Starting A
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskA': finished task run for task with final state: 'Success'
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskB': Starting task run...
Starting B
doing something unrelated
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskB': finished task run for task with final state: 'Success'
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskC': Starting task run...
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskC': finished task run for task with final state: 'Failed'
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskD': Starting task run...
[2020-06-12 06:19:52] INFO - prefect.TaskRunner | Task 'TaskD': finished task run for task with final state: 'Mapped'
[2020-06-12 06:19:52] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
```

Note that TaskC is failing and not mapping even though it is not dumping an error.

I hope I'm not making some silly mistake here again - thanks for the help!

**Comment by joshmeek** [comment]
Just to jump in, @jacques- I am running that code above on master without any errors. Not sure if it would have something to do with it but could you show me the output you get from running `prefect diagnostics` from the command line?

**Comment by jacques-** [comment]
That's weird, I hope it isn't something funny in my system, I'm running in a clean venv so shouldn't be any old/weird stuff in my python - but yeah absolutely - this is when testing on 0.11.5:

```
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "Darwin-19.5.0-x86_64-i386-64bit",
    "prefect_version": "0.11.5",
    "python_version": "3.7.4"
  }
}
```

Are you getting errors on 0.11.5?

And for master I get this exception:

```
> prefect diagnostics
Traceback (most recent call last):
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/bin/prefect", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3191, in <module>
    @_call_aside
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3175, in _call_aside
    f(*args, **kwargs)
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3204, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/Users/jacques/Push/prefect-upstream-tasks-bug/.venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'fsspec>=0.6.0; extra == "bag"' distribution was not found and is required by dask
```

**Comment by joshmeek** [comment]
Ah okay yes I get that failure on `0.11.5` but things work fine for me on master if I refactor the flow a bit:

```python
from prefect import unmapped

with Flow("dummy_flow") as flow:
    task_c_ref = task_c.map(data=task_a, upstream_tasks=[unmapped(task_b)])
    task_d.map(data=task_a, upstream_tasks=[unmapped(task_c_ref)])
```

What's happening is your `.map` call is still attempting to map over the tasks you don't want to map over. Either you should wrap it in `unmapped` or do something more explicit like this:

```python
with Flow("test") as f:
    task_c.set_upstream(task_a, key="data", mapped=True)
    task_c.set_upstream(task_b)

    task_d.set_upstream(task_a, key="data", mapped=True)
    task_d.set_upstream(task_c)
```

We have an idiom for this kind of use case: [Use task mapping to map over a specific set of arguments](https://docs.prefect.io/core/idioms/mapping.html#use-task-mapping-to-map-over-a-specific-set-of-arguments)

**Comment by jacques-** [comment]
Ok, I think I'm following this now. Using imperative API has solved #2753 as well.
 
I think we are going to be converting the rest of our flows to imperative API.

Thanks so much for taking the time to walk me through this, it is really appreciated!

P.S. I see now that https://docs.prefect.io/core/concepts/flows.html#apis talks about the more fine grained control you get from the imperative API - I didn't catch why/when that would be useful when I read those docs originally but def do now!

**Comment by joshmeek** [comment]
@jacques- Glad it's solved! Don't hesitate to reach out with issues here or more questions on Slack if something comes up.
